ouroboros-network-framework/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -103,29 +105,44 @@ dnsResolve :: forall a m s.
     -> DnsSubscriptionTarget
     -> m (SubscriptionTarget m Socket.SockAddr)
 dnsResolve tracer getSeed withResolver peerStatesVar beforeConnect (DnsSubscriptionTarget domain _ _) = do
-    rs <- getSeed
-    withResolver rs $ \resolver -> do
-        ipv6Rsps <- newEmptyTMVarM
-        ipv4Rsps <- newEmptyTMVarM
-        gotIpv6Rsp <- newTVarM False
-
-        -- Though the DNS lib does have its own timeouts, these do not work
-        -- on Windows reliably so as a workaround we add an extra layer
-        -- of timeout on the outside.
-        -- TODO: Fix upstream dns lib.
-        --       On windows the aid_ipv6 and aid_ipv4 threads are leaked incase
-        --       of an exception in the main thread.
-        res <- timeout 20 $ do
-                 aid_ipv6 <- async $ resolveAAAA resolver gotIpv6Rsp ipv6Rsps
-                 aid_ipv4 <- async $ resolveA resolver gotIpv6Rsp ipv4Rsps
-                 rd_e <- waitEitherCatch aid_ipv6 aid_ipv4
-                 handleResult ipv6Rsps ipv4Rsps rd_e
-        case res of
-          Nothing -> do
-            -- TODO: the thread timedout, we should trace it
-            return (SubscriptionTarget $ pure Nothing)
-          Just st ->
-            return st
+    rs_e <- try getSeed
+    case rs_e of
+         Left (e::SomeException) -> do
+             traceWith tracer $ DnsTraceLookupException e
+             -- On windows initializing the DNS library fails with BadConfiguration if the network is down
+             case (fromException e) :: Maybe DNS.DNSError of
+                  Just eb -> return $ listSubscriptionTarget []
+                  Nothing ->
+                    case fromException e :: Maybe IOError of
+                      -- Initializing the DNS library can fail because /etc/resolv.conf is missing. This
+                      -- will happen on OSX if all wifi and eth devices are down.
+                      -- TODO: Do we need to check for the exact IOError type here?
+                      Just eio -> return $ listSubscriptionTarget []
```

Comment:
It's presumably the "does not exist" error, so you can use
```isDoesNotExistError :: IOError -> Bool```

This cascade of checking for different errors might be easier with `catches`
http://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#v:catches
oh but we don't currently provide this in the `MonadCatch` class. We could do though.

Reply:
I know it is possible to use `isDoesNotExistError` to get the exact error. But I'm worried that there may be another case where `makeResolvSeed` throws a different type of IO error and we'll run into the same bug again.

Reply:
I've rewritten it using `catches`.